sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -326,9 +326,9 @@ object ColumnPruning extends Rule[LogicalPlan] {
     case p @ Project(_, a: Aggregate) if (a.outputSet -- p.references).nonEmpty =>
       p.copy(
         child = a.copy(aggregateExpressions = a.aggregateExpressions.filter(p.references.contains)))
-    case p @ Project(_, w: Window) if (w.outputSet -- p.references).nonEmpty =>
+    case p @ Project(_, w: Window)
+        if (AttributeSet(w.windowExpressions.map(_.toAttribute)) -- p.references).nonEmpty =>
```

Comment:
the condition looks pretty complex now, how about we combine this case and the last case(`case p @ Project(_, child) =>`)? e.g.:
```
case p @ Project(_, w: Window) if (w.outputSet -- p.references).nonEmpty =>
val newWindowExprs = w.windowExpressions.filter(p.references.contains)
val newGrandChild = prunedChild(child, AttributeSet(newWindowExprs.flatMap(_.references)))
p.copy(child = w.copy(
  windowExpressions = newWindowExprs,
  child = newGrandChild))
```